torch/csrc/autograd/record_function.h

```diff
@@ -262,52 +347,100 @@ TORCH_API void TEST_unsetGlobalSamplingProbability();
+ * with addThreadLocalCallback
+ */
+TORCH_API bool hasThreadLocalCallbacks();
+
 /**
- * pushCallback adds a pair of callbacks to run with RecordFunction:
- *  start, end - the callbacks to run when entering and exiting the scope;
- *    if start callback returns false, end callback won't be executed;
- *  needs_inputs - whether the callbacks need the inputs passed from the observed
- *    function/range; NOTE: passing the inputs incurs an additional overhead;
- *  sampling_prob - whether the callbacks are sampled and the sampling
- *    probability;
- *  scopes - types of scopes to execute the callbacks on (see RecordScope);
- *    passing empty set means the callbacks will be executed for all possible
- *    scope types
+ * clearThreadLocalCallbacks removes all thread local callbacks
+ */
+TORCH_API void clearThreadLocalCallbacks();
+
+/**
+ * addGlobalCallback adds a global callback to run with RecordFunction:
  *
- * WARNING: not thread safe, must not overlap with other PyTorch code execution
+ * WARNING: not thread safe, typically addGlobalCallback can be called
+ * only during the program initialization
  */
-TORCH_API void pushCallback(
-    std::function<bool(const RecordFunction&)> start,
-    std::function<void(const RecordFunction&)> end =
-        [](const RecordFunction&) {},
-    bool needs_inputs = false,
-    double sampling_prob = 1.0,
-    std::unordered_set<RecordScope, std::hash<RecordScope>> scopes =
-        std::unordered_set<RecordScope, std::hash<RecordScope>>());
+TORCH_API uint64_t addGlobalCallback(RecordFunctionCallback cb);
 
 /**
- * popCallback removes the last pair of callbacks previously added with
- *  pushCallback
+ * removeCallback removes a callback given the handle returned by
+ * addThreadLocalCallback or addGlobalCallback;
  *
- * WARNING: not thread safe, must not overlap with other PyTorch code execution
+ * WARNING: removing a global callback is not thread safe,
+ * no other code can run simultaneously
  */
-TORCH_API void popCallback();
+TORCH_API void removeCallback(uint64_t handle);
 
-// Enable observers thread locally
-TORCH_API void enableObservers(bool enable = true);
+/**
+ * hasGlobalCallbacks returns whether there're global callbacks
+ * registered with pushGlobalCallback
+ */
+TORCH_API bool hasGlobalCallbacks();
 
-// Returns whether observers are enabled (thread locally)
-TORCH_API bool observersEnabled();
+/**
+ * clearGlobalCallbacks removes all global callbacks
+ * WARNING: not thread safe
+ */
+TORCH_API void clearGlobalCallbacks();
+
+// for both thread local and global callbacks
+TORCH_API bool hasCallbacks();
+TORCH_API void clearCallbacks(); // not thread safe
+
+/**
+ * enableRecordFunction enables RecordFunction thread locally
+ */
+TORCH_API void enableRecordFunction(bool enable = true);
+
+/**
+ * isRecordFunctionEnabled returns whether RecordFunction
+ * is enabled thread locally
+ */
+TORCH_API bool isRecordFunctionEnabled();
```

Comment:
is it a temporary solution until we make recordfunction not depend on the key? If so - I'd add a comment on what will be changed in the future

Reply:
no, this is to stay - we do want to disable record functions sometimes, e.g. within record function callbacks themselves

when we move away from TLS dispatcher key, I add another TLS flag (but it's true by default)